Check intersection of lazy images from "update the rendering"

zcorpan opened the pull request:
Fixes #5236.

<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.
-->

- [x] At least two implementers are interested (and none opposed):
   * Chromium: @domfarolino 
   * Gecko: @emilio 
   * WebKit: asked in https://github.com/whatwg/html/pull/5510#issuecomment-625801184
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * N/A per https://github.com/whatwg/html/pull/5510#pullrequestreview-407497015
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/master/MAINTAINERS.md#handling-pull-requests) are filed:
   * N/A per https://github.com/whatwg/html/pull/5510#pullrequestreview-407497015

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/5510/images.html" title="Last updated on May 19, 2020, 9:29 AM UTC (8a764c7)">/images.html</a>  ( <a href="https://whatpr.org/html/5510/caebee3...8a764c7/images.html" title="Last updated on May 19, 2020, 9:29 AM UTC (8a764c7)">diff</a> )
<a href="https://whatpr.org/html/5510/infrastructure.html" title="Last updated on May 19, 2020, 9:29 AM UTC (8a764c7)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/5510/caebee3...8a764c7/infrastructure.html" title="Last updated on May 19, 2020, 9:29 AM UTC (8a764c7)">diff</a> )
<a href="https://whatpr.org/html/5510/urls-and-fetching.html" title="Last updated on May 19, 2020, 9:29 AM UTC (8a764c7)">/urls-and-fetching.html</a>  ( <a href="https://whatpr.org/html/5510/caebee3...8a764c7/urls-and-fetching.html" title="Last updated on May 19, 2020, 9:29 AM UTC (8a764c7)">diff</a> )

zcorpan commented on source:
```diff
@@ -27471,6 +27466,45 @@ was an English &lt;a href="/wiki/Music_hall">music hall&lt;/a> singer, ...</code
+      <p class="note">This allows for fetching the image during scrolling, when it does not, but is
+      about to intersect the viewport.</p>
+     </li>
+     <li><p>Otherwise, <span class="XXX">observe <var>img</var> for intersection</span>.</p></li>
```
This needs to be fleshed out. The `root` needs to be set to something; see https://github.com/whatwg/html/issues/5408#issuecomment-623981827

zcorpan commented on source:
```diff
@@ -27471,6 +27466,45 @@ was an English &lt;a href="/wiki/Music_hall">music hall&lt;/a> singer, ...</code
+  in the <span>run the lazy load intersection steps</span>, which is called from the <span>update
+  the rendering</span> steps.</p>
+
+  <p>The <dfn>run the lazy load intersection steps</dfn> steps for a <code>Document</code> are as
```
It's referenced here https://github.com/whatwg/html/pull/5510/files#diff-36cd38f49b9afa08222c0dc9ebfe35ebR89362-R89364

domfarolino replied:
That link just seems to take me to the top of this page, I'm not sure hwat to look for. When I CTRL+F "run the lazy load intersection steps" here, I see two references:

<img width="742" alt="Screen Shot 2020-05-06 at 12 04 40" src="https://user-images.githubusercontent.com/9669289/81200306-05dacd00-8f81-11ea-8af5-056a42f97deb.png">

I'm guessing there is a third somewhere in the _update the rendering_ steps, that invokes this algorithm? If so, that's what I can't find

zcorpan replied:
Hmm, the page doesn't scroll for some reason, but the lines are highlighted. The reason you don't find it with CTRL+F is that there's a line break. It's at the end of the diff. Search for "the lazy load intersection steps"

domfarolino replied:
Thanks! I was convinced that Chrome's CTRL+F considered line breaks for some reason... I guess not always

zcorpan commented:
Per https://github.com/whatwg/html/issues/5236#issuecomment-591409073 , my understanding was that we only want to start observing in the update the rendering. But maybe I was overthinking it. Should the "update the image data" add lazy images to an intersection observer directly?

domfarolino commented:
> Per #5236 (comment) , my understanding was that we only want to start observing in the update the rendering.

I think from _updating-the-image-data_ if `loading=lazy`, we always want to observe the image, so that the _update-the-rendering_ resumes the _updating-the-image-data_ algorithm.

I'm commenting on how we observe the image, not so much when. My understanding from #5236 is that we don't want to create a per-document image list, and add a new step to _update-the-rendering_. Instead, from _updating-the-image-data_, we want to create and attach an Intersection Observer, and not add anything to the _updating-the-image-data_. That way the next time _update-the-rendering_ runs, we just delegate to the Intersection Observer steps, instead of our own custom logic. Does that make sense? Please see @domenic's comment: https://github.com/whatwg/html/issues/5236#issuecomment-582071666

I'm happy to take on this approach, as it is pretty complicated and interesting to me. Or if you want to adjust this PR to do so, that's fine too.

zcorpan commented:
Thanks for the explanation! It makes sense to me now. I'll update the PR.

zcorpan commented:
OK, please take a look.

I set the `root` to the document, and left `rootMargin` and `threshold` UA-defined.

I realized now though that IntersectionObserver can only initialize `rootMargin`, and then it's read-only. Per https://github.com/whatwg/html/issues/5408#issuecomment-623683336 , I think it needs to be changeable somehow?

domfarolino commented on source:
```diff
@@ -6906,6 +6910,117 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+  a user agent needs to call this algorithm from steps running <span>in parallel</span>, it needs to
+  <span data-x="queue a task">queue</span> a task to do so.</p>
+
+  <p>Each <code>img</code> element have a boolean flag <dfn>ready to be lazy loaded</dfn>, initially
```
s/have/has (the comment I had from earlier)

zcorpan replied:
Oops, thanks for pointing this out again. Fixed now.

zcorpan commented on source:
```diff
@@ -6906,6 +6910,117 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+         <li><p>Set <var>entry</var>'s <span>ready to be lazy loaded</span> flag to true.</p></li>
+
+         <li><p>Call <var>observer</var>'s <code
+         data-x="dom-IntersectionObserver-unobserve">unobserve</code> method with <var>entry</var>
```
Yeah, I know calling methods directly isn't what we want, but IntersectionObserver spec doesn't provide anything else right now. We should file an issue about this.

zcorpan replied:
https://github.com/w3c/IntersectionObserver/issues/427

domfarolino replied:
Thanks for filing

zcorpan commented on source:
```diff
@@ -6906,6 +6910,117 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+   <li><p>Return false.</p></li>
+  </ol>
+
+  <p class="note">This algorithm cannot be called from steps running <span>in parallel</span>. If
```
Fixed.

zcorpan commented:
> I realized now though that IntersectionObserver can only initialize `rootMargin`, and then it's read-only. Per [#5408 (comment)](https://github.com/whatwg/html/issues/5408#issuecomment-623683336) , I think it needs to be changeable somehow?

Filed https://github.com/w3c/IntersectionObserver/issues/428

zcorpan commented:
cc @rwlbuis @smfr

zcorpan commented on source:
```diff
@@ -6906,6 +6910,113 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+        <p>For each <var>entry</var> in <var>entries</var>:
+
+        <ol>
+         <li><p>Set <var>entry</var>'s <span>ready to be lazy loaded</span> flag to true.</p></li>
```
First need to check if the *entry*'s `isIntersecting` is true.

Then set the flag on *entry*'s `target`, not on *entry*

zcorpan commented:
> There are no implementation bugs to be filed, since we're just aligning the spec to what implementations are already doing, so that can probably be checked

http://software.hixie.ch/utilities/js/live-dom-viewer/saved/8073

It seems in both Chromium and Gecko, `root` is the implicit root. Is that what we want? This PR sets it to the image's node document. Using the implicit root takes away the rootMargin if the origins aren't similar-origin, per IntersectionObserver spec (for images in nested browsing contexts).

emilio commented:
Yes, that's what you want since otherwise you load images inside iframes that are out of view.

domfarolino commented on source:
```diff
@@ -6906,6 +6910,113 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+       </dd>
+
+       <dt>threshold</dt>
+       <dd>UA-defined.</dd>
```
I understand the motivation for supplying a `rootMargin`, is supplying a `threshold` useful?

rwlbuis replied:
Same here, I would expect one wants to just use the default (0) for threshold.

rwlbuis commented on source:
```diff
@@ -6906,6 +6910,113 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+         <li><p>Set <var>entry</var>'s <span>ready to be lazy loaded</span> flag to true.</p></li>
+
+         <li><p>Call <var>observer</var>'s <code
+         data-x="dom-IntersectionObserver-unobserve">unobserve</code> method with <var>entry</var>
```
Is this unobserve call needed given "stop intersection-observing a lazy loading element" also calls unobserve?

domfarolino replied:
Agreed, I don't think this is necessary.

domfarolino commented:
I think I've addressed all major reviews in this with my latest commit, would appreciate if another owner took a look (@annevk ?)

zcorpan commented:
Thanks @domfarolino, LGTM

zcorpan commented on source:
```diff
@@ -27692,11 +27776,18 @@ was an English &lt;a href="/wiki/Music_hall">music hall&lt;/a> singer, ...</code
      <li><p>Continue running this algorithm <span>in parallel</span>.</p></li>
 
-     <li><p>Wait until the <span>will lazy load image steps</span> no longer return true, given the
+     <li><p>Wait until the <code>img</code>'s <span>ready to be lazy loaded</span> flag is
```
Sounds good! Can you file a new issue for this?

domfarolino replied:
https://github.com/whatwg/html/issues/5547

annevk commented on source:
```diff
@@ -6906,6 +6912,106 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+   </li>
+
+   <li>
+    <p>If <var>img</var>'s <span>lazy loading attribute</span> is in the <span
```
Use `<li><p>` if `<li>` contains a single `<p>` (equivalent when closing)

annevk commented on source:
```diff
@@ -6906,6 +6912,106 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+   <li><p>Return false.</p></li>
+  </ol>
+
+  <p>Each <code>img</code> element has a boolean flag <dfn>ready to be lazy loaded</dfn>, initially
```
has an associated* and we no longer use flag

annevk commented on source:
```diff
@@ -6906,6 +6912,106 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+
+   <li>
+    <p>If <var>doc</var>'s <span>lazy load intersection observer</span> is null, set it to a new
+    <code>IntersectionObserver</code> instance, initialized as follows:</p>
```
This is quite sketchy. What if JavaScript overwrote this? I think we need a low-level hook in the observer specification for this.

domfarolino replied:
Agreed, see comment below

domenic commented:
> we should not add new points where we instantiate JavaScript objects or call their methods.

Although I agree in general, I'm not sure whether we should block this work on that. It seems like it would require a large refactoring of the IntersectionObserver spec.

A middle ground might be to add awkward disclaimers everywhere, e.g. "new IntersectionObserver instance (using the original value of the IntersectionObserver constructor)", "For each entry in entries (using a method of iteration which does not trigger developer-modifiable array accessors or iteration hooks)", "if entry.isIntersecting is true (using the original value of the isIntersecting getter)", etc.

But, I guess if we could pull off the refactoring, that would be much nicer...

domfarolino commented:
I agree with the comments around using the DOM-exposed method names and objects. This was originally brought up in https://github.com/whatwg/html/pull/5510#discussion_r421524176, and we filed https://github.com/w3c/IntersectionObserver/issues/427 for it.

But yeah, until that gets addressed I agree with making notes in the spec that explain what we're trying to do, and pointing to that issue to track the IO refactoring. For now I've added one around using the `IntersectionObserver` constructor, and I'll add more around the other parts @domenic mentioned.

domfarolino commented:
I've added some more `<p class="XXX">`'s describing that dom-IntersectionObserver* references should use the original values etc. I think this is ready for review

domenic commented on source:
```diff
@@ -6906,6 +6912,135 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
   data-x="invalid value default">invalid value default</i> are both the <span
   data-x="attr-loading-eager-state">Eager</span> state.</p>
 
+  <h5>Processing model for lazy loading attributes</h5>
```
Nit: I might separate this with a `<hr>` instead of having a separate subsection. There's no hard rule but single-subsections always feel a bit weird to me.

domenic commented on source:
```diff
@@ -6906,6 +6912,135 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
   data-x="attr-loading-eager-state">Eager</span> state.</p>
 
+  <h5>Processing model for lazy loading attributes</h5>
+
```
Should this section be phrased in terms of `img` elements, or elements in general? Since "Lazy loading attributes" are currently general.

Right now it seems like a mix.

domfarolino replied:
SGTM

domenic commented on source:
```diff
@@ -6906,6 +6912,135 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+    <code>IntersectionObserver</code> instance, initialized as follows:</p>
+
+    <p class="XXX">The intention is to use the original value of the
+    <code>IntersectionObserver</code> constructor, however we're forced to use the
```
```suggestion
    <code>IntersectionObserver</code> constructor. However, we're forced to use the
```

domenic commented on source:
```diff
@@ -6906,6 +6912,135 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+
+    <p class="XXX">The intention is to use the original value of the
+    <code>IntersectionObserver</code> constructor, however we're forced to use the
+    JavaScript-exposed constructor in this specification, until IntersectionObserver exposes
```
```suggestion
    JavaScript-exposed constructor in this specification, until <cite>Intersection Observer</cite> exposes
```

domenic commented on source:
```diff
@@ -6906,6 +6912,135 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+
+      <ol>
+       <li>
+        <p>For each <var>entry</var> in <var>entries</var> (using a method of iteration which does
+        not trigger developer-modifiable array accessors or iteration hooks):
```
```suggestion
        <p>For each <var>entry</var> in <var>entries</var> <span class="XXX">using a method of iteration which does
        not trigger developer-modifiable array accessors or iteration hooks</span>:
```

(may need re-wrapping)

domenic commented on source:
```diff
@@ -6906,6 +6912,135 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+       <dt>rootMargin</dt>
+       <dd>
+        <p>UA-defined. <span class="XXX">See <a
+        href="https://github.com/whatwg/html/issues/5408">whatwg/html issue #5408</a>.</span></p>
```
```suggestion
        href="https://github.com/whatwg/html/issues/5408">issue #5408</a>.</span></p>
```

domenic commented on source:
```diff
@@ -6906,6 +6912,135 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+        href="https://github.com/whatwg/html/issues/5408">whatwg/html issue #5408</a>.</span></p>
+
+        <p class="note">This allows for fetching the image during scrolling, when it does not, but
+        is about to intersect the viewport.</p>
```
Maybe

> ...when it does not yet — but is about to — intersect the viewport

?

domenic commented on source:
```diff
@@ -6906,6 +6912,135 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+      <dl>
+       <dt>rootMargin</dt>
+       <dd>
+        <p>UA-defined. <span class="XXX">See <a
```
```suggestion
        <p><span>implementation-defined</span>. <span class="XXX">See <a
```

domenic commented on source:
```diff
@@ -6906,6 +6912,135 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+
+     <li>
+      <p>The <var>options</var> is an <code>IntersectionObserverInit</code> dictionary with the
+      following dictionary members:</p>
```
Nit: the `<dl>` structure here feels very heavy. I'd prefer using literal syntax, i.e. `«[ "rootMargin" → an implementation-defined value ]»`

zcorpan commented:
Thanks, @domenic! I've addressed your comments.

Follow-up cleanup tasks will be:
* Fix https://github.com/whatwg/html/issues/5547
* Fix https://github.com/w3c/IntersectionObserver/issues/427 and fix the "XXX" issues introduced in this PR.

zcorpan commented:
@domenic can you do the honors? 🙂
